### PR TITLE
Non-standard objects can be sorted

### PR DESCRIPTION
**Pull request by Schottkyc137**
see issue: #688 #684 

### Visible/Frontend Changes
- [x] Notes can be sorted by category
- [x] Children in schools can be sorted by any value 

### Architectural/Backend Changes
- [x] Any object that has a label will use this label for sorting purposes in an entity-list component
- [x] Added sorting-capabilities to the children-overview component

### CODE REVIEW

File: src/app/child-dev-project/schools/children-overview/children-overview.component.spec.ts [hunk 1]
```diff
@@ -54,4 +54,40 @@ describe("ChildrenOverviewComponent", () => {
     tick();
     expect(component.studentsDataSource.data).toEqual([child1, child2]);
   }));
+
+  function sortAccordingTo(specifier: string, sortedValues: any[]) {
```
Comment (sleidig) [inline src/app/child-dev-project/schools/children-overview/children-overview.component.spec.ts:58]:
```suggestion
  function expectToBeSortedAccordingTo(specifier: string, sortedValues: any[]) {
```

File: src/app/child-dev-project/schools/children-overview/children-overview.component.spec.ts [hunk 2]
```diff
@@ -54,4 +54,40 @@ describe("ChildrenOverviewComponent", () => {
...
+      .sortData(children, component.sort)
+      .map((it) => Number(it.getId()));
+    expect(sortedChildIds).toEqual(ids);
+  }
+
+  it("should sort the table according to the different column values", () => {
+    component.ngAfterViewInit();
+    sortAccordingTo("name", ["AA", "AB", "F", "ZZ"]);
+    sortAccordingTo("projectNumber", [1, 3, 5, 10]);
+    sortAccordingTo("schoolClass", ["AA", "FG", "FH", "I"]);
```
Comment (sleidig) [inline src/app/child-dev-project/schools/children-overview/children-overview.component.spec.ts:90]:
```suggestion
    expectToBeSortedAccordingTo("name", ["AA", "AB", "F", "ZZ"]);
    expectToBeSortedAccordingTo("projectNumber", [1, 3, 5, 10]);
    expectToBeSortedAccordingTo("schoolClass", ["AA", "FG", "FH", "I"]);
```

File: src/app/core/entity-components/entity-list/entity-list.component.spec.ts [hunk 3]
```diff
@@ -256,4 +257,39 @@ describe("EntityListComponent", () => {
...
+    notes[2].category = { id: "2", label: "Z" };
+    notes[1].category = { id: "3", label: "C" };
+    component.ngOnInit();
+    component.entityList = notes;
+    component.sort.sort({ id: "category", start: "asc", disableClear: false });
+    const sortedIds = component.entityDataSource
+      .sortData(notes, component.sort)
+      .map((note) => note.getId());
+    console.log(sortedIds);
+    expect(sortedIds).toEqual(["0", "3", "1", "2"]);
```
Comment (sleidig) [inline src/app/core/entity-components/entity-list/entity-list.component.spec.ts:293]: maybe we could make this test code reusable. I was looking into "parametrized tests" recently: https://stackoverflow.com/questions/5061512/whats-a-good-way-to-reuse-test-code-using-jasmine

File: src/app/core/entity-components/entity-list/entity-list.component.ts [hunk 4]
```diff
@@ -148,6 +148,22 @@ export class EntityListComponent<T extends Entity>
   ngAfterViewInit() {
     this.entityDataSource.sort = this.sort;
     this.entityDataSource.paginator = this.paginator;
+    // sort data according to it's label, if the data has a label
+    // (which it has when using configuration enum types)
+    // otherwise sort by default
+    this.entityDataSource.sortingDataAccessor = (
```
Comment (sleidig) [inline src/app/core/entity-components/entity-list/entity-list.component.ts:154]: @TheSlimvReal can this replace some other workaround we had setting additional fields on the entities for sorting, I think?

File: src/app/core/entity-components/entity-list/entity-list.component.ts [hunk 5]
```diff
@@ -148,6 +148,22 @@ export class EntityListComponent<T extends Entity>
   ngAfterViewInit() {
     this.entityDataSource.sort = this.sort;
     this.entityDataSource.paginator = this.paginator;
+    // sort data according to it's label, if the data has a label
+    // (which it has when using configuration enum types)
+    // otherwise sort by default
```
Comment (sleidig) [inline src/app/core/entity-components/entity-list/entity-list.component.ts:153]: nice solution! :-)

**Review by sleidig — APPROVED** [review]
Looks good :+1:

File: src/app/core/entity-components/entity-list/entity-list.component.ts (hunk 4 above)
Comment (TheSlimvReal) [reply src/app/core/entity-components/entity-list/entity-list.component.ts:154]: Unfortunately not, the problem there was that the data was only loaded, once that row was actually rendered in the list. Therefore, for some entries the attribute was set and for others not, which resulted in wrong sorting behavior.

File: src/app/core/entity-components/entity-list/entity-list.component.spec.ts (hunk 3 above)
Comment (Schottkyc137) [reply src/app/core/entity-components/entity-list/entity-list.component.spec.ts:293]: In general, I think we could profit from some test reusability. For example, it often occurs that one needs an array of entities with certain attributes set (e.g. an array of Notes that all have a certain title). This yields the same 5-10 lines of code in a lot of places. Maybe we should consider refactoring some of the tests to make them look more convenient as well as make further test more easy?

Comment (sleidig) [reply src/app/core/entity-components/entity-list/entity-list.component.spec.ts:293]: yeah, absolutely agree.  Will merge this now to clear the path for the big Angular upgrade but we should look into test code improvements

**Pull request merged**